From ab5fac95f519851122d2446253b27393dac81e7b Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:31:03 -0400
Subject: [scsi] hpsa: only allow REQ_TYPE_FS to use fast path

Message-id: <1410186734-19986-13-git-send-email-thenzl@redhat.com>
Patchwork-id: 92101
O-Subject: [RHEL6.6 e-stor PATCH 12/83] hpsa: only allow REQ_TYPE_FS to use fast path
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

When commands sent down the "fast path" fail, they must be re-tried down the
normal RAID path.  We do this by kicking i/o's back to the scsi mid layer with
a DID_SOFT_ERROR status, which causes them to be retried.  This won't work for
SG_IO's and other non REQ_TYPE_FS i/o's which could get kicked all the way back
to the application, which may have no idea that the command needs resubmitting
and likely no way to resubmit it in such a way the that driver can recognize it
as a resubmit and send it down the normal RAID path.  So we just always send
non REQ_TYPE_FS i/o's down the normal RAID path, never down the "fast path".

Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
(cherry picked from commit 2f6ae5cd24491647a011aead90d47523d875e443)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index 29fbd3a..7d4c45a 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -2348,7 +2348,8 @@ static int hpsa_scsi_queue_command_lck(struct scsi_cmnd *cmd,
  if ((likely(h->transMethod & CFGTBL_Trans_io_accel1)) &&
   (dev->ioaccel_handle) &&
   ((cmd->cmnd[0] == READ_10) || (cmd->cmnd[0] == WRITE_10)) &&
-  (scsi_sg_count(cmd) <= IOACCEL1_MAXSGENTRIES))
+  (scsi_sg_count(cmd) <= IOACCEL1_MAXSGENTRIES) &&
+  likely(cmd->request->cmd_type == REQ_TYPE_FS))
   return hpsa_scsi_ioaccel_queue_command(h, c);
 
  c->Header.ReplyQueue = 0;  /* unused in simple mode */
-- 
1.7.1